### PR TITLE
&&: Kate Pond

### PR DESCRIPTION
**Pull request by Oh-KPond**
# Ride Share
Congratulations! You're submitting your assignment!

I will attempt to answer this questions during a break. I understand that my code is imcomplete.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In the project readme, it asked us to create helper methods in passenger and driver for adding a new trip to the list of trips. However, there was already an add trip method for each of those so I decided to reuse that code instead of creating new. See lines 113 & 114 of lib/trip_dispatcher.rb.
Describe any examples of composition that you encountered in this project, if any | To request a new trip, both the driver and passenger classes needed to be accessed by the trip_dispatcher.
Describe a _nominal test_ that you wrote for this assignment. | In request_trip(passenger_id), @dispatcher.request_trip(5).must_be_kind_of RideShare::Trip.  (In previous projects I almost always forgot to write the ".must_be_kind_of" test. So, I'm glad I did it here.)
Describe an _edge case test_ that you wrote for this assignment. | In initialize for a trip, it should raise an error if end_time is before start_time.
Describe a concept that you gained more clarity on as you worked on this assignment. | It was during group review actually. I had noticed that there were a number of methods that had very similar code, and although I didn't have time to make the code DRY, I asked Charles about it and he explained using inharentence to DRY up the code. It was good to see how that was done. I feel that the practice making new tests was great as well. 

I wish I hadn't been so distracted by my life outside ADA and has finished this project, however with Hotel starting, I feel like I need to move on to remain caught up.

### CODE REVIEW

File: specs/trip_dispatch_spec.rb [hunk 1]
```diff
@@ -89,4 +92,64 @@
...
+      # Then
+      @dispatcher.trips.length.must_equal 601
+    end
+
+    it "adds trip to drivers trips" do
+      # Given
+      # When
+      @dispatcher.request_trip(5)
+      # Then
+      @dispatcher.request_trip(5).driver.trips.length.must_equal 9
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:152]: This test is failing.  

Instead this test should look like:
```ruby
  it "adds trip to drivers trips" do
      # Given
      # When
      trip = @dispatcher.request_trip(5)
      # Then

      trip.driver.trips.length.must_equal 9
    end
```

Notice you're calling `request_trip` twice.   Make sure you're not submitting failing tests in the future.

File: lib/driver.rb [hunk 2]
```diff
@@ -43,5 +45,32 @@ def add_trip(trip)
 
       @trips << trip
     end
+
+    def get_total_rev
+      total_rev = 0
+      @trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:51]: Think about using `inject` or `sum` in the future.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -90,6 +94,33 @@ def load_trips
       trips
     end
 
+    def request_trip(passenger_id)
+      passenger = find_passenger(passenger_id)
+      driver = find_trip_driver(:AVAILABLE)
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:99]: What if no drivers are available?

What if the passenger can't be found.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -77,4 +81,74 @@
       driver.average_rating.must_equal 0
     end
   end
+
+    #For a given driver, calculate their total revenue for all trips. Each driver gets 80% of the trip cost after a fee of $1.65 is subtracted.
+
+  describe "get_total_rev" do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:87]: For these three tests you also need to test'em with no trips.

File: specs/passenger_spec.rb [hunk 5]
```diff
@@ -75,4 +79,63 @@
       end
     end
   end
+
+  describe "total amount of money spent" do
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:83]: Similar here

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -75,4 +79,63 @@
...
+      trip2 = RideShare::Trip.new({id: 8, driver: @driver, passenger: @passenger, start_time: start_time2, end_time: end_time2, cost: 8.00, rating: 5})
+
+      @passenger.add_trip(trip)
+      @passenger.add_trip(trip2)
+      # When
+      # Then
+      @passenger.tot_time_spent.must_equal 3000 # seconds
+    end
+
+    it "returns 0 if no trips have been taken" do
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:134]: 👍

File: specs/spec_helper.rb [hunk 7]
```diff
@@ -1,8 +1,10 @@
 require 'time'
+require 'simplecov'
+SimpleCov.start
```
Comment (CheezItMan) [inline specs/spec_helper.rb:3]: 🥇

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -89,4 +92,64 @@
...
+      # Arrange/Given <= arranged by csv
+      dispatcher = RideShare::TripDispatcher.new
+      # Act/When
+      # Assert/Then
+      dispatcher.find_trip_driver(:AVAILABLE).must_be_kind_of RideShare::Driver
+      dispatcher.find_trip_driver(:AVAILABLE).id.must_equal 2
+    end
+  end
+
+  describe "request_trip(passenger_id)" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:107]: You also need to test that a trip gets added to the passenger, and how it responds if no driver is available and if the passenger id is wrong.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and pretty good commit messages. 
Answer comprehension questions | Check 
**Wave 1** |
`Trip` has a helper method to calculate duration | Check
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue |  Check
Complete tests | Check
**Wave 2** |
`TripDispatcher` has a new method to create trips | Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | **Nope**
Complete tests | Check, see my notes
**Overall** | You definitely have some work to do, but you hit most of the learning goals here.  The biggest thing to work to improve on is looking for edge cases in testing.
